### PR TITLE
Unify go and typescript activation flows.

### PR DESCRIPTION
**Pull request by efritz**
Update some differences between Go and TypeScript extensions:

- Add a max repo limit to TypeScript xref operations (configured in user settings). Already exited in the Go extension.
- Move initLSP function into shared/activate.ts, which was basically the same between the two.
- Added a "show/hide external refs" for TypeScript LSP provider. This feature exists in the Go version and was easy to add to the TypeScript version (also useful as xrefs take a hell of a long time when you need to install all of npm).

### CODE REVIEW

File: shared/activate.ts [hunk 1]
```diff
@@ -79,3 +204,104 @@ function activateWithoutLSP(
...
+        }
+
+        throw err
+    }
+}
+
+/**
+ * When the current settings change, determine if we need to supply/revoke the
+ * externalReferencesProvider to the LSP client. This will cause the LSP client
+ * features to re-register a the references via the extension context.
```
Comment (creachadair) [inline shared/activate.ts]:
```suggestion
 * features to re-register all the references via the extension context.
```

(I think)

File: shared/activate.ts [hunk 2]
```diff
@@ -79,3 +204,104 @@ function activateWithoutLSP(
...
+ * @param implementationId The identifier of the registered locations provider.
+ */
+function registerImplementationsPanel(
+    ctx: sourcegraph.ExtensionContext,
+    implementationId: string
+): void {
+    const panelView = sourcegraph.app.createPanelView(implementationId)
+    panelView.title = 'Implementations'
+    panelView.component = { locationProvider: implementationId }
+    panelView.priority = 160
```
Comment (creachadair) [inline shared/activate.ts]: Why 160?

**Review by creachadair — APPROVED** [review]
I have a few questions and suggestions, but nothing blocking.

File: extensions/typescript/src/settings.ts [hunk 3]
```diff
@@ -73,4 +73,9 @@ export interface Settings {
         | 'normal'
         | 'requestTime'
         | 'verbose'
+
+    /**
+     * The maximum number of dependent packages to look in when searching for external references for a symbol (defaults to 20).
```
[the remark this answers is not in the document]
Comment (efritz) [reply extensions/typescript/src/settings.ts]: This file is generated and formatted by prettier.

File: shared/activate.ts [hunk 4]
```diff
@@ -49,6 +104,76 @@ export async function activateCodeIntel(
     }
 }
 
+/**
+ * Create an LSP client and register providers with the given extension context.
+ * This function returns true if providers are registered and false otherwise.
```
[the remark this answers is not in the document]
Comment (efritz) [reply shared/activate.ts]: That's a good point. This is the behavior matches what's currently there and we should work on a smoother non-LSP code intel fallback on init error. Could you create an issue for this?

File: shared/activate.ts (hunk 1 above)
Comment (efritz) [reply shared/activate.ts]: Oops. Fixed.

File: shared/activate.ts (hunk 2 above)
Comment (efritz) [reply shared/activate.ts]: That's what was there, man! I'm scared to change it.

File: shared/activate.ts [hunk 5]
```diff
@@ -13,6 +22,52 @@ export type LSPFactory = (
...
 ) => Promise<boolean>
 
+/**
+ * A function that creates an LSP client. This function returns the client and
+ * the a features option subject that can be passed feature options at runtime
+ * to change the behavior of the registered providers.
+ *
+ * @param args Parameter bag.
+ */
+export type RegisterClient<S> = (args: {
```
[the remark this answers is not in the document]
Comment (efritz) [reply shared/activate.ts]: Renamed.

File: shared/activate.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (creachadair) [reply shared/activate.ts]: https://github.com/sourcegraph/sourcegraph-basic-code-intel/issues/253

**Pull request merged**